### PR TITLE
refactor(chore) Add mypy static type checking

### LINKED ISSUE #314: Add mypy static type checking
Let's add a static type checking tool to this project and get all of the classes and methods to be statically typed. We can use [mypy](https://mypy-lang.org/) for this since this is what we are mainly using at Mozilla.

- [ ] Add pyright to poetry and configure it in `pyproject.toml`
- [ ] Fix all objects and methods to be type checked
- [ ] Update test files to be type checked
- [ ] Add a command to the `Makefile` to run the static type checking
- [ ] Add a test to the github actions workflow to run the type checker on pull requests. 

Here are some basic options we should use my mypy, we can adjust these later as well. These should exist within the `pyproject.toml` file.

```
[tool.mypy]
python_version = "3.12"
disable_error_code = "attr-defined"
disallow_untyped_calls = false
follow_imports = "normal"
ignore_missing_imports = true
pretty = true
show_error_codes = true
strict_optional = true
warn_no_return = true
warn_redundant_casts = true
warn_return_any = true
warn_unused_ignores = true
warn_unreachable = true
```

### PR DESCRIPTION
**Pull request by cashall-0**
Because

* We want to enhance the project to use static typing

This commit

* Fixes all objects and methods to be type-checked
* Update test files to be type-checked
* Add a command to the Makefile to run the static type checking
* Add a test to the github actions workflow to run the type checker on pull requests.

Fixes #314

### CODE REVIEW

**Comment by cashall-0** [comment]
Hi @b4handjr , the CI type checking is failing because it seems not to be installing mypy. I added mypy to the group dev dependencies, is this appropriate?

File: Makefile [hunk 1]
```diff
@@ -18,3 +18,6 @@ test: install_dependencies
 
 lint: install_dependencies
 	poetry run $(BLACK_CHECK)
+
+typecheck:
+	poetry run mypy .
```
Comment (b4handjr) [inline Makefile:23]: Looks like you forgot to add `install_dependencies` here. Look at the lines above for a clue.

Comment (cashall-0) [reply Makefile]: Thanks, it now installs mypy by adding `install_dependencies`, I was missing that.

**Comment by cashall-0** [comment]
Hi @b4handjr, please can you look into my pr to recommend necessary changes

**Comment by b4handjr** [comment]
Looking at the build on github actions, it seems there is a failing type check in the `docs/` directory, can you please set the type checker to ignore that directory? I also see some failures within `tests/webserver.py`.

**Comment by cashall-0** [comment]
> Looking at the build on github actions, it seems there is a failing type check in the `docs/` directory, can you please set the type checker to ignore that directory? I also see some failures within `tests/webserver.py`.

I have set mypy to ignore the doc folder, and also cleared the webserver error by removing the deprecated import.

**Comment by b4handjr** [comment]
Hello @cashall-0, please pull main and update accordingly, I merge the fix for the failing tests. Thanks!

**Comment by cashall-0** [comment]
> Hello @cashall-0, please pull main and update accordingly, I merge the fix for the failing tests. Thanks!

Alright, I will get to it.

**Comment by b4handjr** [comment]
> > Hello @cashall-0, please pull main and update accordingly, I merge the fix for the failing tests. Thanks!
> 
> I have pulled and updated accordingly, do let me know whatever needs to change as you review.

Thanks! I'll give it a look today

File: foxpuppet/windows/browser/notifications/addons.py [hunk 2]
```diff
@@ -21,7 +22,7 @@ class AddOnInstallConfirmation(BaseNotification):
     """Add-on install confirmation notification."""
 
     @property
-    def addon_name(self):
+    def addon_name(self) -> str | Any:
```
Comment (b4handjr) [inline foxpuppet/windows/browser/notifications/addons.py:25]: Why is this also `Any`?

File: foxpuppet/windows/manager.py [hunk 3]
```diff
@@ -22,8 +24,10 @@ def __init__(self, selenium):
         """
         self.selenium = selenium
 
+    from foxpuppet.windows import BrowserWindow
+
```
Comment (b4handjr) [inline foxpuppet/windows/manager.py:28]: Could this import not go at the top?

File: tests/conftest.py [hunk 4]
```diff
@@ -39,13 +42,13 @@ def webserver():
...
 @pytest.fixture
-def browser(foxpuppet):
+def browser(foxpuppet: FoxPuppet) -> BrowserWindow:
     """First Firefox browser window opened."""
     return foxpuppet.browser
 
 
 @pytest.fixture
-def firefox_options(firefox_options):
+def firefox_options(firefox_options: Any) -> Any:
```
Comment (b4handjr) [inline tests/conftest.py:51]: You can use this, `from selenium.webdriver.firefox.options import Options as FirefoxOptions`

File: tests/test_notifications.py [hunk 5]
```diff
@@ -80,7 +92,9 @@ def complete_notification(browser, confirmation_notification):
     return browser.wait_for_notification(AddOnInstallComplete)
 
 
-def test_open_close_notification(browser, blocked_notification):
+def test_open_close_notification(
+    browser: BrowserWindow, blocked_notification: AddOnInstallBlocked
+) -> BaseNotification:
```
Comment (b4handjr) [inline tests/test_notifications.py:97]: This is a test, it shouldn't return anything

File: tests/webserver.py [hunk 6]
```diff
@@ -32,7 +33,7 @@ def __init__(self, host="", port=8000):
         self.thread = threading.Thread(target=self.server.serve_forever, daemon=True)
 
     @property
-    def host(self):
+    def host(self) -> Any | str:
```
Comment (b4handjr) [inline tests/webserver.py:36]: Is this really Any or string? Why not just a string or int?

Comment (cashall-0) [reply tests/webserver.py:36]: it has the types "str | bytes | bytearray" leaving it at "str" or "str | int" would show an error. But on tracing the variables and method calls it shows type assigned as "_RetAddress" which was used as a type alias for "Any".

File: foxpuppet/expected.py [hunk 7]
```diff
@@ -27,7 +29,10 @@ def __init__(self, selenium, handles):
         self.selenium = selenium
         self.handles = handles
 
-    def __call__(self, *args, **kwargs):
+    if TYPE_CHECKING:
+        from foxpuppet.windows import BrowserWindow  # Import for static typing
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/expected.py:33]: It keeps showing errors of cyclic dependency, like two classes importing from each other. so I imported for just the typechecking.
The current import is done inside the method.
example:
ImportError: cannot import name 'BrowserWindow' from partially initialized module 'foxpuppet.windows' (most likely due to a circular import) (/root/outreachy/moxilla/FoxPuppet/foxpuppet/windows/__init__.py)

File: tests/test_notifications.py [hunk 8]
```diff
@@ -42,8 +45,13 @@ class AddOn(object):
     return AddOn()
 
 
+from .webserver import WebServer
+
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply tests/test_notifications.py:50]: moved to the top and changed to `from tests.webserver import WebServer`

File: tests/webserver.py (hunk 6 above)
Comment (b4handjr) [reply tests/webserver.py:36]: Ah okay gotcha, thanks for that explanation.

File: foxpuppet/windows/manager.py (hunk 3 above)
Comment (cashall-0) [reply foxpuppet/windows/manager.py]: yes, Done

File: foxpuppet/windows/browser/notifications/addons.py (hunk 2 above)
Comment (cashall-0) [reply foxpuppet/windows/browser/notifications/addons.py]: done, changed to str

File: foxpuppet/expected.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (b4handjr) [reply foxpuppet/expected.py:33]: I did some research on this and this pattern seems to be the accepted way

File: tests/test_notifications.py [hunk 9]
```diff
@@ -94,46 +107,60 @@ def test_open_close_notification(browser, blocked_notification):
...
+) -> None:
     """Cancel add-on installation."""
     confirmation_notification.cancel()
     browser.wait_for_notification(None)
 
 
-def test_confirm_addon_install(addon, browser, confirmation_notification):
+def test_confirm_addon_install(
+    addon, browser: BrowserWindow, confirmation_notification: AddOnInstallConfirmation
+) -> None:
```
Comment (b4handjr) [inline tests/test_notifications.py:154]: `addon` should have a type.

File: tests/test_notifications.py [hunk 10]
```diff
@@ -94,46 +107,60 @@ def test_open_close_notification(browser, blocked_notification):
...
     """Confirm add-on installation."""
     assert confirmation_notification.addon_name == addon.name
     confirmation_notification.install()
     browser.wait_for_notification(AddOnInstallComplete)
 
 
-def test_addon_install_complete(addon, browser, complete_notification):
+def test_addon_install_complete(
+    addon, browser: BrowserWindow, complete_notification: AddOnInstallComplete
+) -> None:
```
Comment (b4handjr) [inline tests/test_notifications.py:163]: `addon` should have a type.

File: tests/conftest.py [hunk 11]
```diff
@@ -28,7 +32,7 @@ def pytest_runtest_makereport(item, call):
 
 
 @pytest.fixture(scope="session")
-def webserver():
+def webserver() -> Any:
     """Fixture that starts a local web server."""
     from .webserver import WebServer
```
Comment (b4handjr) [inline tests/conftest.py:37]: This should be able to return something other than `Any`, as it is a yield fixture

**Review by b4handjr — CHANGES_REQUESTED** [review]
Left some of my thoughts. There are also some files that are missing a new line

File: tests/conftest.py (hunk 11 above)
Comment (cashall-0) [reply tests/conftest.py:37]: done used the type `Generator[WebServer, None, None]`

File: tests/test_notifications.py (hunk 10 above)
Comment (cashall-0) [reply tests/test_notifications.py:163]: done

File: tests/test_notifications.py (hunk 9 above)
Comment (cashall-0) [reply tests/test_notifications.py:154]: done

File: tests/test_notifications.py [hunk 12]
```diff
@@ -44,7 +49,9 @@ class AddOn(object):
 
 
 @pytest.fixture
-def blocked_notification(addon, browser, webserver, selenium):
+def blocked_notification(
+    addon: Any, browser: BrowserWindow, webserver: WebServer, selenium: WebDriver
+) -> BaseNotification:
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply tests/test_notifications.py:54]: done, unbundled the class `AddOn` which was previously encapsulated in a method.

File: tests/conftest.py [hunk 13]
```diff
@@ -39,23 +43,23 @@ def webserver():
...
 
 
 @pytest.fixture
-def firefox_options(firefox_options):
+def firefox_options(firefox_options: FirefoxOptions) -> Any:
     """Fixture for configuring Firefox."""
     if os.getenv("MOZREGRESSION_BINARY"):
-        firefox_options.binary = os.getenv("MOZREGRESSION_BINARY")
-    firefox_options.log.level = "trace"
+        firefox_options.binary = os.getenv("MOZREGRESSION_BINARY")  # type: ignore
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply tests/conftest.py:55]: Okay, I have removed the check and binary assignment.

**Comment by cashall-0** [comment]
> Thanks for working on this. I left some comments, I also noticed some of your new files don't have new lines at the end of them.

I might be confused about what new files you are referring to, can you specify one of such?

File: foxpuppet/windows/base.py [hunk 14]
```diff
@@ -26,7 +29,7 @@ def __init__(self, selenium, handle):
         self.wait = WebDriverWait(self.selenium, timeout=10)
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/windows/base.py:29]: it can,I  assigned type `WebDriverWait`

**Comment by cashall-0** [comment]
@b4handjr  I'm trying to update my branch and push but there is a conflict in poetry.lock, the hash generated is different due to changes in `project.toml`. Which of these should I accept? my hash or the hash from main ? or is there a better way to resolve this?

**Comment by b4handjr** [comment]
> @b4handjr I'm trying to update my branch and push but there is a conflict in poetry.lock, the hash generated is different due to changes in `project.toml`. Which of these should I accept? my hash or the hash from main ? or is there a better way to resolve this?

You could delete this file and merge in the new `pyproject.toml` file and just generate a new `poetry.lock` file, that should resolve the conflicts.

File: .github/workflows/linux_tests.yml [hunk 15]
```diff
@@ -50,3 +50,16 @@ jobs:
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.12'
+      - name: Run Type Check
+        shell: bash
+        run: make install_poetry typecheck
```
Comment (b4handjr) [inline .github/workflows/linux_tests.yml:65]: Missing a new line

File: Makefile [hunk 16]
```diff
@@ -18,3 +18,6 @@ test: install_dependencies
 
 lint: install_dependencies
 	poetry run $(BLACK_CHECK)
+
+typecheck: install_dependencies
+	poetry run mypy .
```
Comment (b4handjr) [inline Makefile:23]: Missing a new line

**Review by b4handjr — CHANGES_REQUESTED** [review]

**Comment by cashall-0** [comment]
> > @b4handjr I'm trying to update my branch and push but there is a conflict in poetry.lock, the hash generated is different due to changes in `project.toml`. Which of these should I accept? my hash or the hash from main ? or is there a better way to resolve this?
> 
> You could delete this file and merge in the new `pyproject.toml` file and just generate a new `poetry.lock` file, that should resolve the conflicts.
> 
> Let me know when you get that resolved and I'll take another look. It's looking good so far, thank you!

While trying to resolve this, I noticed poetry version changed from `1.8.4` to `1.8.3` on the main. Is there a particular reason for the downgrade?

**Comment by cashall-0** [comment]
@b4handjr , I have tried to resolve the conflict and fixed from your recommendations. you can now take another look , thanks.

**Comment by b4handjr** [comment]
> @b4handjr , I have tried to resolve the conflict and fixed from your recommendations. you can now take another look , thanks.

Thanks, I'll take a look but I think this is looking good

File: tests/webserver.py [hunk 17]
```diff
@@ -32,7 +33,7 @@ def __init__(self, host="", port=8000):
         self.thread = threading.Thread(target=self.server.serve_forever, daemon=True)
```
Comment (b4handjr) [inline tests/webserver.py:33]: This could use a type

**Review by b4handjr — CHANGES_REQUESTED** [review]
A few spots in the main modules that are missing types. I don't think we need to be too picky in the tests. Thanks for the work so far, this is just about done!

File: foxpuppet/windows/browser/notifications/base.py [hunk 18]
```diff
@@ -31,15 +37,15 @@ def create(window, root):
             :py:class:`BaseNotification`: Firefox notification.
 
         """
-        notifications = {}
+        notifications: dict[Any, Any] = {}
         _id = root.get_property("id")
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/windows/browser/notifications/base.py]: Done, it is of the type `str` but mypy wouldn't let my assign only that so I used suggested `str | bool | WebElement | dict`

File: foxpuppet/windows/browser/notifications/addons.py [hunk 19]
```diff
@@ -32,12 +33,12 @@ def addon_name(self):
             el = self.find_description()
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/windows/browser/notifications/addons.py]: done. el is of type `webElement`

File: foxpuppet/windows/base.py [hunk 20]
```diff
@@ -37,21 +40,21 @@ def document_element(self):
...
-    def firefox_version(self):
+    def firefox_version(self) -> int:
         """Major version of Firefox in use.
 
         Returns:
             int: Major component of the Firefox version.
 
         """
-        version = self.selenium.capabilities["browserVersion"]
+        version: Any = self.selenium.capabilities["browserVersion"]
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/windows/base.py]: done, `str`

File: foxpuppet/windows/base.py [hunk 21]
```diff
@@ -23,10 +26,10 @@ def __init__(self, selenium, handle):
         """
         self.selenium = selenium
         self.handle = handle
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/windows/base.py]: done.

File: tests/webserver.py (hunk 17 above)
Comment (cashall-0) [reply tests/webserver.py]: done.

File: foxpuppet/windows/browser/notifications/addons.py [hunk 22]
```diff
@@ -6,12 +6,14 @@
 from selenium.webdriver.common.by import By
 
 from foxpuppet.windows.browser.notifications import BaseNotification
+from typing import Any
```
Comment (b4handjr) [inline foxpuppet/windows/browser/notifications/addons.py:9]: I don't see a use for Any in this file.

**Review by b4handjr — CHANGES_REQUESTED** [review]
Just a few final comments. I'd like to not use `Any` as much as possible, but there will be some places where we need it.

File: foxpuppet/windows/browser/notifications/base.py [hunk 23]
```diff
@@ -60,37 +66,37 @@ def origin(self):
...
+    def find_description(self) -> WebElement | Any:
         """Retrieve the notification description."""
         if self.window.firefox_version >= 67:
             return self.root.find_element(By.CLASS_NAME, "popup-notification-description")
         return self.root.find_anonymous_element_by_attribute(
             "class", "popup-notification-description"
         )
 
-    def find_close_button(self):
+    def find_close_button(self) -> WebElement | Any:
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/windows/browser/notifications/base.py:93]: Yes , but for this method and four others written this way
`def find_close_button(self) -> WebElement | Any:
        """Retrieve the close button."""
        if self.window.firefox_version >= 67:
            return self.root.find_element(By.CLASS_NAME, "popup-notification-closebutton")
        return self.root.find_anonymous_element_by_attribute("anonid", "closebutton")`
        
for the second return statement here, I can't seem to find the implementation of the method `find_anonymous_element_by_attribute()`    hence, the any for whatever the outcome.

File: foxpuppet/windows/base.py [hunk 24]
```diff
@@ -5,14 +5,17 @@
 
 from selenium.webdriver.common.by import By
 from selenium.webdriver.support.wait import WebDriverWait
+from selenium.webdriver.remote.webdriver import WebDriver
+from selenium.webdriver.remote.webelement import WebElement
+from typing import Any
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/windows/base.py:10]: removed

File: tests/test_notifications.py [hunk 25]
```diff
@@ -80,7 +96,9 @@ def complete_notification(browser, confirmation_notification):
     return browser.wait_for_notification(AddOnInstallComplete)
 
 
-def test_open_close_notification(browser, blocked_notification):
+def test_open_close_notification(
+    browser: BrowserWindow, blocked_notification: AddOnInstallBlocked
+):
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply tests/test_notifications.py:101]: done

File: tests/test_notifications.py [hunk 26]
```diff
@@ -94,46 +112,63 @@ def test_open_close_notification(browser, blocked_notification):
         (AddOnInstallBlocked, "AddOnInstallBlocked was not shown"),
     ],
 )
-def test_wait_for_notification_timeout(browser, _class, message):
+def test_wait_for_notification_timeout(browser: BrowserWindow, _class: Any, message: Any):
```
[the remark this answers is not in the document]
Comment (cashall-0) [reply tests/test_notifications.py:115]: done

File: foxpuppet/windows/browser/notifications/addons.py (hunk 22 above)
Comment (cashall-0) [reply foxpuppet/windows/browser/notifications/addons.py]: cleared

File: foxpuppet/windows/browser/notifications/base.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (cashall-0) [reply foxpuppet/windows/browser/notifications/base.py:93]: @b4handjr can you look at this?

**Review by b4handjr — APPROVED** [review]
Thanks so much for this @cashall-0 ! This looks great and will help us as we expand this tool. Awesome work 🎉 🎉 🎉 🎉

**Pull request merged**